### ISSUE TITLE
useEffect or useLayoutEffect

### ISSUE THREAD

**Issue by dai-shi** [avkonst/hookstate#40]
Let's move the discussion about this topic from https://github.com/avkonst/hookstate/issues/14#issuecomment-585547432

**Comment by dai-shi** [comment]
> Well, I actually wanted to use useEffect everywhere and drop useLayoutEffect.

Yeah, we should always prefer using useEffect.

> wondered how "safe" it would be for Hookstate

I'm not sure as I don't look into the recent hookstate code.
Updating refs in useEffect is probably fine for Blocking Mode.
The reason I use useLayoutEffect in react-tracked is for Concurrent Mode. (I have two useLayoutEffects for different reasons: a) to update refs and b) to subscribe. a) is the same in react-redux. b) is for my specific hack in my three libs [[more info about the hack](https://github.com/dai-shi/reactive-react-redux/issues/29)]).

**Comment by avkonst** [comment]
basically, a user interaction or network event CAN NOT TRIGGER any state change BEFORE useEffect is completed but AFTER dom drawing is completed?

**Comment by avkonst** [comment]
@dai-shi right or no? what do you think?

**Comment by avkonst** [comment]
for reference for the future: https://twitter.com/dai_shi/status/1227237850707447809

**Comment by dai-shi** [comment]
https://reactjs.org/docs/hooks-reference.html#useeffect
>Although useEffect is deferred until after the browser has painted, it’s guaranteed to fire before any new renders. React will always flush a previous render’s effects before starting a new update.

You can trigger state updates before effects, but it will not render.
So, React "state" won't change, but any other JS variables can change.

Note in CM, it can render multiple times before effect runs, so you can't reliably mutate variables in render functions.

**Comment by avkonst** [comment]
Nice thanks. It means I can boost the performance a little bit more
